contracts/SharedAsset.sol

```diff
@@ -0,0 +1,866 @@
+        // idx must be a valid index in the list of owners
+        require(idx < ownerList.length, "");
+        return ownerList[idx];
+    }
+
+    // increase the amount of shares for a given address;
+    // if that address was not previously an owner, it becomes one;
+    // proposal support is adjusted based on new amount of shares
+    function increaseShares(address ownerAddress, uint amount) private {
+        Owner storage owner = ownershipMap[ownerAddress];
+        if (owner.shares == 0) {
+            addOwner(ownerAddress, amount);
+        } else {
+            owner.sharesHistory[owner.sharesHistorySize] = DataPoint(owner.shares, block.timestamp);
+            ++owner.sharesHistorySize;
+            owner.pendingPayout += (totalPayout - owner.lastKnownPayout) * owner.shares / TOTAL_SHARES;
+            owner.lastKnownPayout = totalPayout;
+            owner.shares += amount;
+        }
+    }
+
+    // decrease the amount of shares for a given owner;
+    // if the owner is left without shares, they are removed from the list of owners;
+    // proposal support is adjusted based on new amount of shares
+    function decreaseShares(address ownerAddress, uint amount) private {
+        Owner storage owner = ownershipMap[ownerAddress];
+        owner.sharesHistory[owner.sharesHistorySize] = DataPoint(owner.shares, block.timestamp);
+        ++owner.sharesHistorySize;
+        owner.pendingPayout += (totalPayout - owner.lastKnownPayout) * owner.shares / TOTAL_SHARES;
+        owner.lastKnownPayout = totalPayout;
+        owner.shares -= amount;
+    }
+
+    // transfers shares from one account to another
+    function transferShares(address from, address to, uint sharesAmount) private {
+        decreaseShares(from, sharesAmount);
+        increaseShares(to, sharesAmount);
+    }
+
+    // creates a new owner
+    function addOwner(address ownerAddress, uint shares) private {
+        ownershipMap[ownerAddress] = Owner({
+            shares: shares,
+            sharesOnSale: 0,
+            listPosition: ownerList.length,
+            supportedProposals: new uint[](0),
+            lastKnownPayout : totalPayout,
+            pendingPayout : 0,
+            sharesHistorySize : 0
+        });
+        ownerList.push(ownerAddress);
+    }
+
+    // removes an owner from the ownerList
+    function removeOwner(address ownerAddress) private {
+        uint pos = ownershipMap[ownerAddress].listPosition;
+        ownerList[pos] = ownerList[ownerList.length - 1];
+        ownershipMap[ownerList[pos]].listPosition = pos;
+        --ownerList.length;
+    }
+
+    
+
+    //*****************************PAYMENTS****************************************************
+
+    // the ether produced by the asset(s) is sent by calling this function
+    function payment(string info) external payable {
+        require(treasuryRatio <= TREASURY_RATIO_DENOMINATOR, "Payments are not allowed to invalid contracts.");
```

Comment:
This check should not occur here, it occurs when proposing to change the reasuryRatio. Also, follow the style of the other require statements, these error messages eat available bytecode.

Reply:
Yes, the check has to happen when changing the treasuryRatio somewhere else in the code, but that doesn't mean this code shouldnt check for this, otherwise it will still process payments on an invalid configuration and send out potentially too much money.

Reply:
an invalid configuration should not happen. if there is a way it can happen it needs to be fixed. no need for redundancy checks.